File: pages/refactorings/replace-mutable-fields-with-lenses/03-file-tree/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

genName :: Gen String
genName = ("n" <>) . show <$> Gen.int (Range.linear 0 9)

genSize :: Gen Int
genSize = Gen.int (Range.linear (-20) 20)

genNode :: Int -> Gen Before.Node
genNode depth =
  let leaf = Before.Node <$> genName <*> genSize <*> pure []
  in if depth == 0 then leaf
     else Gen.choice
       [ leaf
       , Before.Node <$> genName <*> genSize
           <*> Gen.list (Range.linear 0 3) (genNode (depth - 1))
       ]

toAfter :: Before.Node -> After.Node
toAfter (Before.Node n s cs) = After.Node n s (map toAfter cs)

-- toAfter maps a whole tree faithfully, so After's result can be
-- compared with the converted Before result.
prop_agrees :: Property
prop_agrees = property $ do
  n  <- forAll (genNode 3)
  by <- forAll genSize
  After.bumpSizes (toAfter n) by === toAfter (Before.bumpSizes n by)

prop_laws :: Property
prop_laws = property $ do
  n  <- forAll (genNode 1)
  v1 <- forAll genSize
  v2 <- forAll genSize
  let l   = After.sizeLens
      m   = toAfter n
  After.get l (After.set l v1 m) === v1
  After.set l (After.get l m) m === m
  After.set l v2 (After.set l v1 m) === After.set l v2 m

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("bumpSizes: Before == After", prop_agrees)
    , ("size lens obeys the laws on generated nodes", prop_laws)
    ]
  unless ok exitFailure

File: pages/refactorings/replace-mutable-fields-with-lenses/03-file-tree/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("bumpSizes: Before == After", agrees),
    property("the size lens obeys get-put, put-get, put-put", laws),
  )

  val genName: Gen[String] =
    for n <- Gen.int(Range.linear(0, 9)) yield "n" + n.toString
  val genSize: Gen[Int] = Gen.int(Range.linear(-20, 20))

  def genNode(depth: Int): Gen[Before.Node] =
    val leaf =
      for n <- genName; s <- genSize yield Before.Node(n, s, Nil)
    if depth == 0 then leaf
    else
      Gen.choice1(
        leaf,
        for
          n  <- genName
          s  <- genSize
          cs <- genNode(depth - 1).list(Range.linear(0, 3))
        yield Before.Node(n, s, cs)
        ,
      )

  def toAfter(n: Before.Node): After.Node = n match
    case Before.Node(nm, s, cs) => After.Node(nm, s, cs.map(toAfter))

  def same(b: Before.Node, a: After.Node): Boolean = (b, a) match
    case (Before.Node(n1, s1, c1), After.Node(n2, s2, c2)) =>
      n1 == n2 && s1 == s2 && c1.size == c2.size &&
        c1.zip(c2).forall((x, y) => same(x, y))

  def agrees: Property =
    for
      n  <- genNode(3).forAll
      by <- genSize.forAll
    yield
      val b = Before.bumpSizes(n, by)
      val a = After.bumpSizes(toAfter(n), by)
      same(b, a) ==== true

  def laws: Property =
    for
      n  <- genNode(1).forAll
      v1 <- genSize.forAll
      v2 <- genSize.forAll
    yield
      val l = After.size
      val m = toAfter(n)
      l.get(l.set(v1)(m)) ==== v1 and
        l.set(l.get(m))(m) ==== m and
        l.set(v2)(l.set(v1)(m)) ==== l.set(v2)(m)

@main def spec(): Unit =
  val results = Props.tests.map { t =>
    val r = Property.check(t.withConfig(PropertyConfig.default),
      t.result, Seed.fromTime())
    println(Test.renderReport(
      "Props", t, r, ansiCodesSupported = false))
    r.status
  }
  if !results.forall(_ == Status.ok) then sys.exit(1)
